core/src/main/scala/chisel3/Mem.scala

```diff
@@ -392,4 +392,272 @@ sealed class SyncReadMem[T <: Data] private[chisel3] (
+    * port directionality based on connection semantics and the `when` context unlike SyncReadMem.apply(),
+    * so the behavior of the port must be controlled by changing the values of the input parameters.
+    *
+    * @param idx memory element index to write into
+    * @param writeData new data to write
+    * @param enable enables access to the memory
+    * @param isWrite performs a write instead of a read when enable is true; the return
+    * value becomes undefined when this parameter is true
+    *
+    * @return The value of the memory at idx when enable is true and isWrite is false,
+    * or an undefined value otherwise.
+    *
+    * @example Controlling a read/write port with IO signals
+    * {{{
+    * class MyMemWrapper extends Module {
+    *   val width = 2
+    *
+    *   val io = IO(new Bundle {
+    *     val address = Input(UInt())
+    *     val wdata = Input(UInt(width.W))
+    *     val enable = Input(Bool())
+    *     val isWrite = Input(Bool())
+    *     val rdata = Output(UInt(width.W))
+    *   })
+    *
+    *   val mem = SyncReadMem(2, UInt(width.W))
+    *   val rwPort = mem.readWrite(io.address, io.wdata, io.enable, io.isWrite)
+    *   io.rdata := rwPort
+    * }
+    *
+    * class Top extends Module {
+    *   val mem = Module(new MyMemWrapper)
+    *   mem.io := DontCare
+    *
+    *   val rdata = Wire(UInt(2.W))
+    *   readData := DontCare
+    *
+    *   // Read from the memory port
+    *   when(doRead) {
+    *     mem.io.enable := true.B
+    *     mem.io.isWrite := false.B
+    *     mem.io.address := someAddress
+    *   }
+    *   // Read data will show up on the next cycle
+    *   when(timeToRead) {
+    *     readData := mem.io.rdata
+    *   }
+    *
+    *   // Write to the memory
+    *   when(doWrite) {
+    *     mem.io.enable := true.B
+    *     mem.io.isWrite := true.B
+    *     mem.io.address := anotherAddress
+    *     mem.io.wdata := 3.U
+    *   }
+    * }
+    * }}}
+    */
+  def readWrite(idx: UInt, writeData: T, en: Bool, isWrite: Bool): T = macro SourceInfoTransform.idxDataEnIswArg
+
+  /** @group SourceInfoTransformMacro */
+  def do_readWrite(idx: UInt, writeData: T, en: Bool, isWrite: Bool)(implicit sourceInfo: SourceInfo): T =
+    _readWrite_impl(idx, writeData, en, isWrite, Builder.forcedClock, true)
+
+  /** Generates an explicit read-write port for this SyncReadMem, using a clock that may be
+    * different from the implicit clock.
+    *
+    * @param idx memory element index to write into
+    * @param writeData new data to write
+    * @param enable enables access to the memory
+    * @param isWrite performs a write instead of a read when enable is true; the return
+    * value becomes undefined when this parameter is true
+    * @param clock clock to bind to this read-write port
+    *
+    * @return The value of the memory at idx when enable is true and isWrite is false,
+    * or an undefined value otherwise.
+    */
```

Comment:
```suggestion
    * @return The read port of the memory, which will return the value of the memory at idx when enable is true and isWrite is false,
    * or an undefined value otherwise, on the next rising clock edge
    */
```